### PR TITLE
Implemented remove_pack() and show_cellular_data_confirmation() and PlayAssetPackException class.

### PR DESCRIPTION
**Pull request by Arivval**
Included relevant Gut unit test cases.

### CODE REVIEW

File: api/src/model/request/play_asset_delivery_request.gd [hunk 1]
```diff
@@ -20,7 +20,7 @@
 # signals emitted from PlayAssetDelivery Android Plugin.
 #
 # ##############################################################################
-class_name PlayAssetPackRequest
+class_name PlayAssetDeliveryRequest
```
Comment (Arivval) [inline api/src/model/request/play_asset_delivery_request.gd:23]: I renamed the base class to PlayAssetPackRequest since we will be having a subclass named PlayAssetPackRequest (Used for ```request_asset_pack()``` and ```get_asset_pack_state()```).

File: api/test/test_helper/fake_android_plugin.gd [hunk 2]
```diff
@@ -127,3 +164,30 @@ func cancel(pack_names : Array):
 			return_asset_pack_states[PlayAssetPackStates._PACK_STATES_KEY][pack_name] = current_asset_pack_dict
 
 	return return_asset_pack_states
+
+# -----------------------------------------------------------------------------
+# Simulates the showCellularDataConfirmation() function in PlayAssetDelivery 
+# Android plugin. Emits signal with arguments configured using 
+# set_show_confirmation_response().
```
Comment (Arivval) [inline api/test/test_helper/fake_android_plugin.gd:171]: Since we need to test exception cases that can not be easily triggered with normally, I used the preconfigured signal approach for these two functions.

File: api/test/test_play_asset_pack_manager.gd [hunk 3]
```diff
@@ -161,3 +162,117 @@ func test_cancel_asset_pack_request_non_existing_pack_name():
...
+	mock_plugin.set_show_confirmation_response(true, PlayAssetPackManager.AssetPackStorageMethod.STORAGE_FILES, {})
+	var test_object = create_play_asset_pack_manager(mock_plugin)
+	
+	var request_object = test_object.show_cellular_data_confirmation()
+	
+	# connect to the helper function, simulating the workflow connecting callback to signal
+	request_object.connect("request_completed", self, "show_cellular_data_confirmation_success_helper")
+	
+	# yield to the request_completed signal for no longer than 1 seconds and assert for signal emitted
+	yield(yield_to(request_object, "request_completed", 1), YIELD)
```
Comment (Arivval) [inline api/test/test_play_asset_pack_manager.gd:179]: This [syntax](https://github.com/bitwes/Gut/wiki/Yielding) for yield is specifically used for Gut testing environment.

File: api/test/test_helper/fake_android_plugin.gd [hunk 4]
```diff
@@ -85,6 +99,29 @@ func remove_asset_pack_state(pack_name : String):
...
+	_remove_pack_success = success
+	_remove_pack_error = error
+
+# -----------------------------------------------------------------------------
+# Helper function that emits signal from another thread with latency so we 
+# have time to connect to that signal on main thread for testing.
+# -----------------------------------------------------------------------------
+func emit_signal_helper(args : Array):
+	# Delay this thread by 100 milliseconds, allowing us to connect/yield to signal in time.
+	OS.delay_msec(100)
```
Comment (Arivval) [inline api/test/test_helper/fake_android_plugin.gd:117]: [OS.delay_msec()](https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-delay-msec)

File: api/test/test_play_asset_pack_manager.gd [hunk 5]
```diff
@@ -161,3 +162,117 @@ func test_cancel_asset_pack_request_non_existing_pack_name():
...
+	
+	# assert using getters, simulating the workflow of yielding the signals
+	assert_true(not request_object.get_status())
+	assert_asset_pack_exception_eq_dict(request_object.get_error(), \
+		create_mock_asset_pack_java_lang_exception_dict())
+	
+	# join instantiated thread
+	mock_plugin._show_confirmation_thread.wait_to_finish()
+
+func show_cellular_data_confirmation_error_helper(status : bool, error : PlayAssetPackException):
```
Comment (jamcohen) [inline api/test/test_play_asset_pack_manager.gd:218]: `assert_show_cellular_data_confirmation_signal_is_error` might be a better name. Something to indicate that this method is there to verify the completed signal arguments.

File: api/test/test_play_asset_pack_manager.gd [hunk 6]
```diff
@@ -161,3 +162,117 @@ func test_cancel_asset_pack_request_non_existing_pack_name():
...
+	yield(yield_to(request_object, "request_completed", 1), YIELD)
+	assert_signal_emitted(request_object, "request_completed", "signal should have emitted")
+	
+	# assert using getters, simulating the workflow of yielding the signals
+	assert_true(request_object.get_status())
+	
+	# join instantiated thread
+	mock_plugin._remove_pack_thread.wait_to_finish()
+
+func remove_pack_success_helper(status : bool, error : PlayAssetPackException):
```
Comment (jamcohen) [inline api/test/test_play_asset_pack_manager.gd:245]: Same as above, see if you can make this name more descriptive.

File: api/test/test_helper/fake_android_plugin.gd [hunk 7]
```diff
@@ -24,11 +24,25 @@
...
+signal removePackError
+signal showCellularDataConfirmationSuccess
+signal showCellularDataConfirmationError
+
 var _asset_location_store : Dictionary
 var _asset_pack_location_store : Dictionary
 
 var _asset_pack_states_store : Dictionary
 
+var _show_confirmation_thread : Thread
```
Comment (jamcohen) [inline api/test/test_helper/fake_android_plugin.gd:37]: This class is starting to have a lot of groups of fields/methods with different prefixes, and I imagine there will be more as we add other methods. Is there a way to break those elements up into their own classes and have fake_android_plugin reference them?

For example these fields/methods could be broken out into another class `fake_cellular_confirmation_handler`:
````
var _show_confirmation_thread : Thread
var _show_confirmation_success : bool
var _show_confirmation_result : int
var _show_confirmation_error : Dictionary
func set_show_confirmation_response()
````

Comment (jamcohen) [reply api/test/test_helper/fake_android_plugin.gd:37]: Feel free to pick a different name than `fake_cellular_confirmation_handler `.

File: api/src/model/request/play_asset_pack_remove_request.gd [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# This object also provides relevant getters so that it is possible to retrieve
+# the updated states from this object using the yield to signal approach.
+#
+# ##############################################################################
+class_name PlayAssetPackRemoveRequest
+extends PlayAssetDeliveryRequest
+
+signal request_completed
+
+var _status : bool
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/model/request/play_asset_pack_remove_request.gd:35]: Refactored ```var _status``` to ```var _did_succeed```.

File: api/src/model/request/play_asset_pack_remove_request.gd [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+#	distributed under the License is distributed on an "AS IS" BASIS,
+#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#	See the License for the specific language governing permissions and
+#	limitations under the License.
+#
+# ##############################################################################
+#
+# Request object that handles asynchronous logic related to remove_pack().
+# 
+# Emits request_completed signal upon success/error. For this signal, the first
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/model/request/play_asset_pack_remove_request.gd:21]: Refactored this comment as a block comment above signal declaration.

File: api/test/test_helper/fake_android_plugin.gd (hunk 7 above)
Comment (jamcohen) [reply api/test/test_helper/fake_android_plugin.gd:37]: btw this refactoring can come in a later PR.

File: api/src/model/request/play_cellular_data_confirmation_request.gd [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+func get_error() -> PlayAssetPackException:
+	return _error
+
+# -----------------------------------------------------------------------------
+# Callback functions handling signals emitted from the plugin.
+# -----------------------------------------------------------------------------
+func on_show_cellular_data_confirmation_success(result : int):
+	_status = true
+	_result = result
+	emit_signal("request_completed", true, result)
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/model/request/play_cellular_data_confirmation_request.gd]: I think we discussed this request_{completed, succeeded, failed} decision during the design doc meeting and removing the request_completed signal will make it hard for devs who are yielding on that signal.

I think refactoring this signal to take three arguments is definitely a good call. I added this change in the most recent commit.
Since for this result the type is int we are unable to return null. Is returning -1 for result a good decision here?

File: api/src/model/request/play_asset_pack_remove_request.gd [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+# PlayAssetPackException object representing the exception encountered.
+#
+# This object also provides relevant getters so that it is possible to retrieve
+# the updated states from this object using the yield to signal approach.
+#
+# ##############################################################################
+class_name PlayAssetPackRemoveRequest
+extends PlayAssetDeliveryRequest
+
+signal request_completed
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/model/request/play_asset_pack_remove_request.gd]: Refactored this and added signal hints:
```
signal request_completed(did_succeed, result, exception)
```
I also tried static typing for these argument names but it is not supported here : /

File: api/src/play_asset_pack_manager.gd [hunk 12]
```diff
@@ -82,6 +94,25 @@ func _initialize_plugin() -> Object:
 		push_error("Android plugin singleton not found!")
 		return null
 
+# -----------------------------------------------------------------------------
+# Helper functions that receive signals emitted from the plugin
+# -----------------------------------------------------------------------------
+func show_cellular_data_confirmation_success(result : int, signal_id : int):
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/play_asset_pack_manager.gd]: Good catch, I added underscore prefix for these functions.

[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/play_asset_pack_manager.gd]: Yeah I think this is a good call. I renamed these functions with _forward prefix.

File: api/src/play_asset_pack_manager.gd [hunk 13]
```diff
@@ -69,7 +69,19 @@ func _ready():
 
 func _initialize():
 	_plugin_singleton = _initialize_plugin()
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/play_asset_pack_manager.gd:71]: Since Godot's Android Plugin also used this [singleton naming approach](https://docs.godotengine.org/en/stable/tutorials/plugins/android/android_plugin.html#loading-and-using-an-android-plugin), I think going with the current naming can be more consistent with Godot's doc.

File: api/test/test_helper/base_test_class.gd [hunk 14]
```diff
@@ -122,3 +128,9 @@ func create_mock_asset_pack_state_with_status_dict(pack_name : String, status :
 		PlayAssetPackState._TRANSFER_PROGRESS_PERCENTAGE_KEY: 0
 	}
 
+func create_mock_asset_pack_java_lang_exception_dict():
+	return {
+		PlayAssetPackException._TYPE_KEY: "java.lang.Exception", 
+		PlayAssetPackException._MESSAGE_KEY: "Just an usual exception",
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/base_test_class.gd:134]: Thanks so much for pointing out this edge case in syntax parsing! Updated this line.

File: api/test/test_helper/fake_android_plugin.gd [hunk 15]
```diff
@@ -24,11 +24,25 @@
 class_name FakeAndroidPlugin
 extends Object
 
+signal removePackSuccess
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd:27]: Added arguments.

File: api/test/test_helper/base_test_class.gd [hunk 16]
```diff
@@ -122,3 +128,9 @@ func create_mock_asset_pack_state_with_status_dict(pack_name : String, status :
 		PlayAssetPackState._TRANSFER_PROGRESS_PERCENTAGE_KEY: 0
 	}
 
+func create_mock_asset_pack_java_lang_exception_dict():
+	return {
+		PlayAssetPackException._TYPE_KEY: "java.lang.Exception", 
+		PlayAssetPackException._MESSAGE_KEY: "Just an usual exception", 
+		PlayAssetPackException._ERROR_CODE_KEY: -100
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/base_test_class.gd]: updated.

File: api/test/test_helper/fake_android_plugin.gd [hunk 17]
```diff
@@ -85,6 +99,29 @@ func remove_asset_pack_state(pack_name : String):
...
+
+func set_remove_pack_response(success : bool, error : Dictionary):
+	_remove_pack_success = success
+	_remove_pack_error = error
+
+# -----------------------------------------------------------------------------
+# Helper function that emits signal from another thread with latency so we 
+# have time to connect to that signal on main thread for testing.
+# -----------------------------------------------------------------------------
+func emit_delayed_signal_helper(args : Array):
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd:115]: Renamed to ```emit_delayed_signal_helper()```

File: api/test/test_helper/fake_android_plugin.gd [hunk 18]
```diff
@@ -127,3 +164,34 @@ func cancel(pack_names : Array):
...
+# Android plugin. Emits signal with arguments configured using 
+# set_show_confirmation_response().
+# -----------------------------------------------------------------------------
+func showCellularDataConfirmation(signal_id : int):
+	# use multithreading to call emit_delayed_signal_helper() to emit signal with delay since Godot's main
+	# thread is blocking
+	_show_confirmation_thread = Thread.new()
+	if _show_confirmation_success:
+		var thread_args = ["showCellularDataConfirmationSuccess", _show_confirmation_result, signal_id]
+		_show_confirmation_thread.start(self, "emit_delayed_signal_helper", thread_args)
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd]: Yup. I just double checked and signals will be received on a new thread regardless of which thread the signal was subscribed on/emitted from.

[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd]: Updated: for both yield and callbacks, the signals are received on the thread the signal was emitted from.

[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd]: TODO: Test for concurrency behavior and which thread is the Android Plugin's signals on. (Also added TODO comment in code base.)

File: api/test/test_helper/fake_android_plugin.gd [hunk 19]
```diff
@@ -127,3 +164,34 @@ func cancel(pack_names : Array):
...
+	else:
+		var thread_args = ["showCellularDataConfirmationError", _show_confirmation_error, signal_id]
+		_show_confirmation_thread.start(self, "emit_delayed_signal_helper", thread_args)
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd]: Refactored this string to const.

File: api/test/test_play_asset_pack_manager.gd [hunk 20]
```diff
@@ -161,3 +162,117 @@ func test_cancel_asset_pack_request_non_existing_pack_name():
...
+	assert_signal_emitted(request_object, "request_completed", "signal should have emitted")
+	
+	# assert using getters, simulating the workflow of yielding the signals
+	assert_true(request_object.get_status())
+	assert_eq(request_object.get_result(), PlayAssetPackManager.AssetPackStorageMethod.STORAGE_FILES)
+	
+	# join instantiated thread
+	mock_plugin._show_confirmation_thread.wait_to_finish()
+
+func show_cellular_data_confirmation_success_helper(status : bool, result : int):
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_play_asset_pack_manager.gd]: Renamed these helper function.

File: api/test/test_play_asset_pack_manager.gd [hunk 21]
```diff
@@ -161,3 +162,117 @@ func test_cancel_asset_pack_request_non_existing_pack_name():
...
+	
+	# connect to helper function, simulating the workflow of connecting callback to signal
+	request_object.connect("request_completed", self, "show_cellular_data_confirmation_success_helper")
+	
+	# yield to the request_completed signal for no longer than 1 seconds and assert for signal emitted
+	yield(yield_to(request_object, "request_completed", 1), YIELD)
+	assert_signal_emitted(request_object, "request_completed", "signal should have emitted")
+	
+	# assert using getters, simulating the workflow of yielding the signals
+	assert_true(request_object.get_status())
```
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_play_asset_pack_manager.gd]: ```show_cellular_data_confirmation_success_helper()``` will be called automatically upon ```emit_signal``` since it is already connected in line 176. The assertions below is also needed since we need to test the workflow for yielding to signals.

File: api/test/test_play_asset_pack_manager.gd (hunk 5 above)
Comment (Arivval) [reply api/test/test_play_asset_pack_manager.gd]: Renamed this method to ```assert_show_cellular_data_confirmation_signal_is_error ```.

File: api/test/test_play_asset_pack_manager.gd (hunk 6 above)
Comment (Arivval) [reply api/test/test_play_asset_pack_manager.gd]: Renamed to ```assert_remove_pack_is_success```

File: api/test/test_play_asset_pack_manager.gd (hunk 20 above)
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_play_asset_pack_manager.gd]: Update: re-renamed this function to ```assert_show_cellular_data_confirmation_is_success``` to make it more consistent with ```assert_show_cellular_data_confirmation_signal_is_error``` naming.

File: api/test/test_helper/fake_android_plugin.gd (hunk 7 above)
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd:39]: Yeah I think refactoring these variables into their own classes is definitely a better design decision. I added a TODO comment for this suggestion and will work on this in next PR.

File: api/src/play_asset_pack_manager.gd (hunk 12 above)
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/play_asset_pack_manager.gd]: I also added underscore for request callback functions such as ```_on_show_cellular_data_confirmation_success``` since developers should not be calling them.

File: api/test/test_play_asset_pack_manager.gd (hunk 21 above)
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_play_asset_pack_manager.gd]: The ```	yield(yield_to(request_object, "request_completed", 1), YIELD) ````statement also asserts the ```request_completed``` must be received and the method connected must be called.

File: api/test/test_helper/fake_android_plugin.gd (hunk 18 above)
[the remark this answers is not in the document]
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd]: Updated code to use ```call_deferred()``` for current PR.

File: api/src/model/request/play_cellular_data_confirmation_request.gd (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arivval) [reply api/src/model/request/play_cellular_data_confirmation_request.gd]: Fixed a bug where the wrong enum is used, create a new 
```
enum CellularDataConfirmationResult {
	RESULT_UNDEFINED = -2,
	RESULT_OK = -1,
	RESULT_CANCELED = 0
}
```
to handle this situation.

File: api/test/test_helper/fake_android_plugin.gd [hunk 22]
```diff
@@ -24,11 +24,28 @@
 class_name FakeAndroidPlugin
 extends Object
 
+signal removePackSuccess(signalID)
```
Comment (jamcohen) [inline api/test/test_helper/fake_android_plugin.gd:27]: Should these be separated by underscores rather than camelCase?

File: api/test/test_helper/fake_android_plugin.gd [hunk 23]
```diff
@@ -127,3 +167,34 @@ func cancel(pack_names : Array):
 			return_asset_pack_states[PlayAssetPackStates._PACK_STATES_KEY][pack_name] = current_asset_pack_dict
 
 	return return_asset_pack_states
+
+# -----------------------------------------------------------------------------
+# Simulates the showCellularDataConfirmation() function in PlayAssetDelivery 
+# Android plugin. Emits signal with arguments configured using 
+# set_show_confirmation_response().
+# -----------------------------------------------------------------------------
+func showCellularDataConfirmation(signal_id : int):
```
Comment (jamcohen) [inline api/test/test_helper/fake_android_plugin.gd:176]: `showCellularDataConfirmation` -> `show_cellular_data_confirmation`

Same with removePack below.

File: api/test/test_helper/fake_android_plugin.gd (hunk 22 above)
Comment (Arivval) [reply api/test/test_helper/fake_android_plugin.gd:27]: This camelCase naming for mock signals/functions in fake_android_plugin is intentional, since we are mocking existing functions and signals written in Java.
